flaky test tools_tests_testosqueryd causes: The operation was canceled

Common Azure output
```
2019-12-19T00:57:05.6482809Z test 67
2019-12-19T00:57:05.6483487Z       Start 67: tools_tests_testosqueryd
2019-12-19T00:57:05.6486937Z 
2019-12-19T00:57:05.6488167Z 67: Test command: "C:\Program Files\CMake\bin\cmake.exe" "-E" "env" "PYTHONPATH=D:/a/1/b/build/python_path" "C:/hostedtoolcache/windows/Python/3.8.0/x64/python.exe" "-u" "test_osqueryd.py" "--verbose" "--build" "D:/a/1/b/build" "--test-configs-dir" "D:/a/1/b/build/test_configs"
2019-12-19T00:57:05.6489417Z 67: Test timeout computed to be: 60
2019-12-19T00:57:07.0327242Z 67: test_1_daemon_without_watchdog (__main__.DaemonTests) ... ok
2019-12-19T00:57:10.0606738Z 67: test_2_daemon_with_option (__main__.DaemonTests) ... ok
2019-12-19T00:57:20.1347596Z 67: test_3_daemon_lost_worker (__main__.DaemonTests) ... ok
2019-12-19T00:57:23.1978638Z 67: test_3_daemon_with_watchdog (__main__.DaemonTests) ... Thrift: Thu Dec 19 00:57:23 2019 TPipeServer ConnectNamedPipe GLE=errno = 995
2019-12-19T00:57:24.1650753Z 67: ok
2019-12-19T00:57:25.1776474Z 67: test_4_daemon_sighup (__main__.DaemonTests) ... ok
2019-12-19T00:57:27.1959503Z 67: test_5_daemon_sigint (__main__.DaemonTests) ... ok
2019-12-19T00:57:30.2344707Z 67: test_6_logger_mode (__main__.DaemonTests) ... ok
2019-12-19T00:57:30.5499932Z 67: test_7_logger_stdout (__main__.DaemonTests) ... severity=0 location=init.cpp:418 message=osquery initialized [version=4.1.2-3-g423ea4bc]
2019-12-19T00:57:30.5502109Z 67: severity=0 location=system.cpp:362 message=Writing osqueryd pid (3524) to C:\Users\VSSADM~1\AppData\Local\Temp\osquery-tests-python-VssAdministrator\tests.pid
2019-12-19T00:57:30.5502897Z 67: severity=0 location=rocksdb.cpp:131 message=Opening RocksDB handle: C:\Users\VSSADM~1\AppData\Local\Temp\osquery-tests-python-VssAdministrator\tests.db5473
2019-12-19T00:57:30.5503566Z 67: severity=0 location=init.cpp:656 message=An error occured during extension manager startup: Extensions disabled
2019-12-19T00:57:30.5504311Z 67: severity=0 location=auto_constructed_tables.cpp:93 message=Removing stale ATC entries
2019-12-19T00:57:30.5505364Z 67: severity=0 location=options.cpp:100 message=Verbose logging enabled by config option
2019-12-19T00:57:30.5510464Z 67: severity=0 location=main.cpp:103 message=Not starting the distributed query service: Distributed query service not enabled.
2019-12-19T00:57:31.9176535Z 67: ok
2019-12-19T00:57:31.9177747Z 67: test_8_hostid_uuid (__main__.DaemonTests) ... severity=0 location=init.cpp:418 message=osquery initialized [version=4.1.2-3-g423ea4bc]
2019-12-19T00:57:31.9179080Z 67: severity=0 location=system.cpp:362 message=Writing osqueryd pid (3916) to C:\Users\VSSADM~1\AppData\Local\Temp\osquery-tests-python-VssAdministrator\tests.pid
2019-12-19T00:57:31.9180107Z 67: severity=0 location=rocksdb.cpp:131 message=Opening RocksDB handle: C:\Users\VSSADM~1\AppData\Local\Temp\osquery-tests-python-VssAdministrator\tests.db6702
2019-12-19T00:57:31.9180665Z 67: severity=0 location=init.cpp:656 message=An error occured during extension manager startup: Extensions disabled
2019-12-19T00:57:31.9180954Z 67: severity=0 location=auto_constructed_tables.cpp:93 message=Removing stale ATC entries
2019-12-19T00:57:31.9181238Z 67: severity=0 location=options.cpp:100 message=Verbose logging enabled by config option
2019-12-19T00:57:31.9181468Z 67: severity=0 location=system.cpp:289 message=Using host identifier: E49BFB5F-4A61-4749-BB8B-D2EF792BC0AF
2019-12-19T00:57:32.2674807Z 67: ok
2019-12-19T00:57:32.5956708Z 67: test_9_hostid_instance (__main__.DaemonTests) ... severity=0 location=init.cpp:418 message=osquery initialized [version=4.1.2-3-g423ea4bc]
2019-12-19T00:57:32.5957722Z 67: severity=0 location=system.cpp:362 message=Writing osqueryd pid (4260) to C:\Users\VSSADM~1\AppData\Local\Temp\osquery-tests-python-VssAdministrator\tests.pid
2019-12-19T00:57:32.5958274Z 67: severity=0 location=rocksdb.cpp:131 message=Opening RocksDB handle: C:\Users\VSSADM~1\AppData\Local\Temp\osquery-tests-python-VssAdministrator\tests.db4945
2019-12-19T00:57:32.5958689Z 67: severity=0 location=init.cpp:656 message=An error occured during extension manager startup: Extensions disabled
2019-12-19T00:57:32.5958879Z 67: severity=0 location=auto_constructed_tables.cpp:93 message=Removing stale ATC entries
2019-12-19T00:57:32.5959571Z 67: severity=0 location=options.cpp:100 message=Verbose logging enabled by config option
2019-12-19T00:57:32.5960149Z 67: severity=0 location=system.cpp:289 message=Using host identifier: 3cacf385-7842-4c6c-b65e-e446f660f2e4
2019-12-19T00:57:33.2820996Z 67: ok
2019-12-19T00:57:33.2821860Z 67: 
2019-12-19T00:57:33.2822123Z 67: ----------------------------------------------------------------------
2019-12-19T00:57:33.2822333Z 67: Ran 10 tests in 27.249s
2019-12-19T00:57:33.2822555Z 67: 
2019-12-19T00:57:33.2822739Z 67: OK
2019-12-19T02:00:39.2773911Z ##[error]The operation was canceled.
2019-12-19T02:00:39.2828032Z ##[section]Finishing: Run tests
```

Notice the time difference between

```
2019-12-19T00:57:33.2822739Z 67: OK
2019-12-19T02:00:39.2773911Z ##[error]The operation was canceled.
```

--- a/osquery/config/tests/config_tests.cpp
+++ b/osquery/config/tests/config_tests.cpp
@@ -37,7 +37,6 @@
 #include <thread>
 #include <vector>
 
-
 namespace osquery {
 
 DECLARE_uint64(config_refresh);
@@ -80,6 +79,13 @@ class ConfigTests : public testing::Test {
   void TearDown() {
     fs::remove_all(fake_directory_);
     FLAGS_config_refresh = refresh_;
+  }
+
+  void resetDispatcher() {
+    auto& dispatcher = Dispatcher::instance();
+    dispatcher.stopServices();
+    dispatcher.joinServices();
+    dispatcher.resetStopping();
   }
 
  protected:
@@ -560,8 +566,7 @@ TEST_F(ConfigTests, test_config_refresh) {
   get().reset();
 
   // Stop the existing refresh runner thread.
-  Dispatcher::stopServices();
-  Dispatcher::joinServices();
+  resetDispatcher();
 
   // Set a config_refresh value to convince the Config to start the thread.
   FLAGS_config_refresh = 2;
@@ -603,8 +608,7 @@ TEST_F(ConfigTests, test_config_refresh) {
   EXPECT_EQ(get().getRefresh(), FLAGS_config_refresh);
 
   // Stop the new refresh runner thread.
-  Dispatcher::stopServices();
-  Dispatcher::joinServices();
+  resetDispatcher();
 
   FLAGS_config_refresh = refresh;
   FLAGS_config_accelerated_refresh = refresh_acceleratred;

--- a/osquery/dispatcher/dispatcher.cpp
+++ b/osquery/dispatcher/dispatcher.cpp
@@ -85,6 +85,11 @@ Status Dispatcher::addService(InternalRunnableRef service) {
   return Status::success();
 }
 
+void Dispatcher::resetStopping() {
+  WriteLock lock(mutex_);
+  stopping_ = false;
+}
+
 void Dispatcher::removeService(const InternalRunnable* service) {
   auto& self = Dispatcher::instance();
   WriteLock lock(self.mutex_);
@@ -137,7 +142,6 @@ void Dispatcher::joinServices() {
     }
   }
 
-  self.stopping_ = false;
   DLOG(INFO) << "Services and threads have been cleared";
 }
 

--- a/osquery/dispatcher/dispatcher.cpp
+++ b/osquery/dispatcher/dispatcher.cpp
@@ -76,8 +76,8 @@ Status Dispatcher::addService(InternalRunnableRef service) {
     auto thread = std::make_unique<std::thread>(
         std::bind(&InternalRunnable::run, &*service));
     DLOG(INFO) << "Adding new service: " << service->name() << " ("
-              << service.get() << ") to thread: " << thread->get_id() << " ("
-              << thread.get() << ") in process " << platformGetPid();
+               << service.get() << ") to thread: " << thread->get_id() << " ("
+               << thread.get() << ") in process " << platformGetPid();
 
     self.service_threads_.push_back(std::move(thread));
     self.services_.push_back(std::move(service));

--- a/osquery/dispatcher/dispatcher.cpp
+++ b/osquery/dispatcher/dispatcher.cpp
@@ -65,20 +65,19 @@ Status Dispatcher::addService(InternalRunnableRef service) {
   }
 
   auto& self = instance();
-  if (self.stopping_) {
-    // Cannot add a service while the dispatcher is stopping and no joins
-    // have been requested.
-    return Status(1, "Cannot add service, dispatcher is stopping");
-  }
-
-  auto thread = std::make_unique<std::thread>(
-      std::bind(&InternalRunnable::run, &*service));
-
-  DLOG(INFO) << "Adding new service: " << service->name() << " ("
-             << service.get() << ") to thread: " << thread->get_id() << " ("
-             << thread.get() << ") in process " << platformGetPid();
   {
     WriteLock lock(self.mutex_);
+    if (self.stopping_) {
+      // Cannot add a service while the dispatcher is stopping and no joins
+      // have been requested.
+      return Status(1, "Cannot add service, dispatcher is stopping");
+    }
+
+    auto thread = std::make_unique<std::thread>(
+        std::bind(&InternalRunnable::run, &*service));
+    DLOG(INFO) << "Adding new service: " << service->name() << " ("
+              << service.get() << ") to thread: " << thread->get_id() << " ("
+              << thread.get() << ") in process " << platformGetPid();
 
     self.service_threads_.push_back(std::move(thread));
     self.services_.push_back(std::move(service));
@@ -144,11 +143,11 @@ void Dispatcher::joinServices() {
 
 void Dispatcher::stopServices() {
   auto& self = instance();
-  self.stopping_ = true;
-
-  WriteLock lock(self.mutex_);
   DLOG(INFO) << "Thread: " << std::this_thread::get_id()
              << " requesting a stop";
+
+  WriteLock lock(self.mutex_);
+  self.stopping_ = true;
   for (const auto& service : self.services_) {
     assureRun(service);
     service->interrupt();

--- a/osquery/extensions/tests/extensions.cpp
+++ b/osquery/extensions/tests/extensions.cpp
@@ -65,11 +65,18 @@ class ExtensionsTest : public testing::Test {
   }
 
   void TearDown() override {
-    Dispatcher::stopServices();
-    Dispatcher::joinServices();
+    resetDispatcher();
+
     if (!isPlatform(PlatformType::TYPE_WINDOWS)) {
       fs::remove(fs::path(socket_path));
     }
+  }
+
+  void resetDispatcher() {
+    auto& dispatcher = Dispatcher::instance();
+    dispatcher.stopServices();
+    dispatcher.joinServices();
+    dispatcher.resetStopping();
   }
 
   bool ping(int attempts = 3) {

--- a/osquery/include/osquery/dispatcher.h
+++ b/osquery/include/osquery/dispatcher.h
@@ -161,9 +161,7 @@ class Dispatcher : private boost::noncopyable {
 
  private:
   /// For testing only, reset the stopping status for unittests.
-  void resetStopping() {
-    stopping_ = false;
-  }
+  void resetStopping();
 
  private:
   /// The set of shared osquery service threads.
@@ -187,11 +185,14 @@ class Dispatcher : private boost::noncopyable {
    * A future join will be requested AFTER all services were expected to have
    * been interrupted.
    */
-  std::atomic<bool> stopping_{false};
+  bool stopping_{false};
 
  private:
   friend class InternalRunnable;
-  friend class ExtensionsTests;
+
+  // Tests
+  friend class ConfigTests;
   friend class DispatcherTests;
+  friend class ExtensionsTest;
 };
 } // namespace osquery

